File: src/main/java/de/twometer/neko/gui/GuiManager.java
package de.twometer.neko.gui;

import com.labymedia.ultralight.*;
import com.labymedia.ultralight.bitmap.UltralightBitmapSurface;
import com.labymedia.ultralight.config.FontHinting;
import com.labymedia.ultralight.config.UltralightConfig;
import com.labymedia.ultralight.math.IntRect;
import com.labymedia.ultralight.plugin.loading.UltralightLoadListener;
import de.twometer.neko.core.NekoApp;
import de.twometer.neko.event.Events;
import de.twometer.neko.event.SizeChangedEvent;
import de.twometer.neko.render.model.Mesh;
import de.twometer.neko.render.model.ModelBase;
import de.twometer.neko.render.shading.IShadingStrategy;
import de.twometer.neko.res.AssetPaths;
import de.twometer.neko.util.CrashHandler;
import de.twometer.neko.util.Log;
import org.greenrobot.eventbus.Subscribe;
import org.joml.Matrix4f;

import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL12.*;

public class GuiManager implements UltralightLoadListener {

    private I18n i18n = new I18n();

    private Matrix4f guiMatrix;

    private UltralightPlatform platform;
    private UltralightRenderer renderer;
    private UltralightView view;

    private int texture = -1;

    private GuiShader guiShader;
    private ModelBase quadModel;

    private boolean wasCursorVisible = false;
    private Page loadingScreen;
    private Page currentPage;

    public void create() {
        Log.d("Initializing gui system");
        i18n.load();

        var ultralightPath = AssetPaths.NATIVE_PATH + "Ultralight/";
        var resourcePath = ultralightPath + "resources/";
        var nativesDir = Path.of(ultralightPath);
        try {
            if (!Files.exists(nativesDir))
                throw new UltralightLoadException("Native ultralight libraries not found. Make sure you have the latest ultralight libraries in " + ultralightPath);

            UltralightJava.extractNativeLibrary(nativesDir);
            UltralightJava.load(nativesDir);
        } catch (UltralightLoadException e) {
            CrashHandler.fatal(e);
        }

        var window = NekoApp.get().getWindow();
        platform = UltralightPlatform.instance();
        platform.setConfig(
                new UltralightConfig()
                        .resourcePath(resourcePath)
                        .fontHinting(FontHinting.NORMAL)
                        .deviceScale(window.getScale())
        );
        platform.usePlatformFileSystem(AssetPaths.GUI_PATH);
        platform.usePlatformFontLoader();
        platform.setLogger(new GuiLogger());
        platform.setClipboard(new GuiClipboard());

        renderer = UltralightRenderer.create();

        view = renderer.createView(window.getWidth(), window.getHeight(), true);
        view.setViewListener(new GuiViewListener());
        view.setLoadListener(this);

        var adapter = new InputAdapter(view);
        adapter.attach();

        quadModel = Mesh.create(4, 2)
                .putVertex(0, 0)
                .putVertex(1, 0)
                .putVertex(0, 1)
                .putVertex(1, 1)
                .bake("GuiQuad", GL_TRIANGLE_STRIP);

        guiShader = NekoApp.get().getShaderProvider().getShader(GuiShader.class);

        Events.register(this);

        Log.i("Gui System initialized");
    }

    public void showPage(Page page) {
        if (currentPage == null) // On open, save cursor state
            wasCursorVisible = NekoApp.get().getWindow().isCursorVisible();

        if (currentPage != null && page == null) {
            currentPage.onUnload();
            currentPage.setContext(null);
        }

        currentPage = page;

        if (currentPage == null) { // On close, restore cursor state
            NekoApp.get().getWindow().setCursorVisible(wasCursorVisible);
        } else {
            showUrl("file:///" + page.getPath().replace("\\", "/"));
            // Apply cursor state
            NekoApp.get().getWindow().setCursorVisible(currentPage.isCursorVisible());
        }
    }

    public void showUrl(String url) {
        Log.i("Navigating to " + url);
        this.view.loadURL(url);
    }

    public void render() {
        if (texture == -1) genTexture();
        if (currentPage == null) return;

        glDisable(GL_DEPTH_TEST);
        glDepthMask(false);

        renderer.update();
        renderer.render();

        // We wanna do our own shading here
        NekoApp.get().getRenderManager().setShadingStrategy(IShadingStrategy.NOP);

        var surface = (UltralightBitmapSurface) view.surface();
        var bitmap = surface.bitmap();

        var width = (int) view.width();
        var height = (int) view.height();

        glBindTexture(GL_TEXTURE_2D, texture);
        IntRect dirtyBounds = surface.dirtyBounds();

        if (dirtyBounds.isValid()) {
            ByteBuffer imageData = bitmap.lockPixels();

            glPixelStorei(GL_UNPACK_ROW_LENGTH, (int) bitmap.rowBytes() / 4);
            if (dirtyBounds.width() == width && dirtyBounds.height() == height) {
                // Update full image
                glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, width, height, 0, GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV, imageData);
                glPixelStorei(GL_UNPACK_ROW_LENGTH, 0);
            } else {
                // Update partial image
                int x = dirtyBounds.x();
                int y = dirtyBounds.y();
                int dirtyWidth = dirtyBounds.width();
                int dirtyHeight = dirtyBounds.height();
                int startOffset = (int) ((y * bitmap.rowBytes()) + x * 4);

                glTexSubImage2D(
                        GL_TEXTURE_2D,
                        0,
                        x, y, dirtyWidth, dirtyHeight,
                        GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV,
                        imageData.position(startOffset));
            }
            glPixelStorei(GL_UNPACK_ROW_LENGTH, 0);

            bitmap.unlockPixels();
            surface.clearDirtyBounds();
        }

        guiShader.bind();
        guiShader.guiMatrix.set(guiMatrix);
        quadModel.render();

        glDepthMask(true);
        glEnable(GL_DEPTH_TEST);
    }

    @Subscribe
    public void onSizeChanged(SizeChangedEvent e) {
        view.resize(e.width, e.height);
        guiMatrix = new Matrix4f().ortho2D(0, 1.0f, 1.0f, 0);
    }

    public I18n getI18n() {
        return i18n;
    }

    private void genTexture() {
        Log.d("Creating GUI texture");
        texture = glGenTextures();
        glBindTexture(GL_TEXTURE_2D, texture);
        glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
        glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
        glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
    }

    @Override
    public void onBeginLoading(long frameId, boolean isMainFrame, String url) {

    }

    @Override
    public void onFinishLoading(long frameId, boolean isMainFrame, String url) {

    }

    @Override
    public void onFailLoading(long frameId, boolean isMainFrame, String url, String description, String errorDomain, int errorCode) {

    }

    @Override
    public void onUpdateHistory() {

    }

    @Override
    public void onWindowObjectReady(long frameId, boolean isMainFrame, String url) {

    }

    @Override
    public void onDOMReady(long frameId, boolean isMainFrame, String url) {
        if (currentPage != null) {
            var pageCtx = new PageContext(view);
            pageCtx.runInJsContext(js -> {
                var databind = new Databind(DatabindConfiguration.builder().contextProviderFactory(pageCtx).build());
                var remote = js.makeObject(databind.toJavascript(currentPage.getClass()), new DatabindJavascriptClass.Data(currentPage, null));
                js.getGlobalContext().getGlobalObject().setProperty("_remote", remote, 0);
            });
            currentPage.setContext(pageCtx);
            currentPage.onDomReady();
            pageCtx.runScript("OnLoaded()");
        }
    }

    public void setLoadingScreen(Page loadingScreen) {
        this.loadingScreen = loadingScreen;
    }

    public Page getLoadingScreen() {
        return loadingScreen;
    }

    public Page getCurrentPage() {
        return currentPage;
    }
}

File: neko-engine/src/main/kotlin/de/twometer/neko/gui/GuiManager.kt
package de.twometer.neko.gui

import com.labymedia.ultralight.UltralightRenderer
import com.labymedia.ultralight.UltralightView
import com.labymedia.ultralight.bitmap.UltralightBitmapSurface
import com.labymedia.ultralight.databind.Databind
import com.labymedia.ultralight.databind.DatabindConfiguration
import com.labymedia.ultralight.javascript.JavascriptEvaluationException
import com.labymedia.ultralight.plugin.loading.UltralightLoadListener
import de.twometer.neko.core.NekoApp
import de.twometer.neko.events.Events
import de.twometer.neko.events.ResizeEvent
import de.twometer.neko.render.OpenGL
import de.twometer.neko.render.Primitives
import de.twometer.neko.render.Shader
import de.twometer.neko.render.Texture2d
import de.twometer.neko.res.ShaderCache
import mu.KotlinLogging
import org.greenrobot.eventbus.Subscribe
import org.lwjgl.opengl.GL12.*
import java.nio.ByteBuffer

private val logger = KotlinLogging.logger {}

class GuiManager : UltralightLoadListener {

    private val globalObjects = HashMap<String, Any>()
    private lateinit var shader: Shader
    private lateinit var renderer: UltralightRenderer
    private lateinit var view: UltralightView
    private lateinit var contextProvider: PageContextProvider

    private val loadLock = Object()
    var finishedLoading: Boolean = false
        private set

    var page: Page? = null
        set(newVal) {
            synchronized(loadLock) {
                if (!finishedLoading && field != null)
                    loadLock.wait()
            }

            field?.onUnloaded()
            field = newVal
            finishedLoading = false
            if (newVal != null)
                view.loadURL("file:///${newVal.path}")
        }

    private val texture: Texture2d by lazy {
        val id = glGenTextures()
        glBindTexture(GL_TEXTURE_2D, id)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE)
        glBindTexture(GL_TEXTURE_2D, 0)
        Texture2d(id, 0, 0)
    }

    fun setup() {
        Events.register(this)
        UltralightLoader.load()
        renderer = UltralightRenderer.create()

        val (width, height) = NekoApp.the.window.getSize()
        view = renderer.createView(width.toLong(), height.toLong(), true)
        view.setViewListener(UltralightNekoViewListener())
        view.setLoadListener(this)
        view.focus()
        Events.register(UltralightNekoInputAdapter(view))

        contextProvider = PageContextProvider(view)
        contextProvider.syncWithJavascript {
            val config = DatabindConfiguration.builder().contextProviderFactory(contextProvider).build()
            contextProvider.databind = Databind(config)
        }

        shader = ShaderCache.get("base/gui.nks")

        logger.info { "GUI system loaded" }
    }

    fun render() {
        if (page == null)
            return

        OpenGL.disable(GL_DEPTH_TEST)
        shader.bind()
        texture.bind()

        renderer.update()
        renderer.render()

        val surface = view.surface() as UltralightBitmapSurface
        val bitmap = surface.bitmap()
        val dirtyBounds = surface.dirtyBounds()
        val (width, height) = view.width().toInt() to view.height().toInt()

        if (dirtyBounds.isValid) {
            val imageData: ByteBuffer = bitmap.lockPixels()
            glPixelStorei(GL_UNPACK_ROW_LENGTH, bitmap.rowBytes().toInt() / 4)
            if (dirtyBounds.width() == width && dirtyBounds.height() == height) {
                // Update full image
                glTexImage2D(
                    GL_TEXTURE_2D,
                    0,
                    GL_RGBA8,
                    width,
                    height,
                    0,
                    GL_BGRA,
                    GL_UNSIGNED_INT_8_8_8_8_REV,
                    imageData
                )
                glPixelStorei(GL_UNPACK_ROW_LENGTH, 0)
            } else {
                // Update partial image
                val x = dirtyBounds.x()
                val y = dirtyBounds.y()
                val dirtyWidth = dirtyBounds.width()
                val dirtyHeight = dirtyBounds.height()
                val startOffset = (y * bitmap.rowBytes() + x * 4).toInt()
                glTexSubImage2D(
                    GL_TEXTURE_2D,
                    0,
                    x, y, dirtyWidth, dirtyHeight,
                    GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV,
                    imageData.position(startOffset)
                )
            }
            glPixelStorei(GL_UNPACK_ROW_LENGTH, 0)
            bitmap.unlockPixels()
            surface.clearDirtyBounds()
        }

        Primitives.unitQuad.render()

        shader.unbind()
        OpenGL.enable(GL_DEPTH_TEST)
    }

    fun runScript(script: String): String {
        return try {
            view.evaluateScript(script)
        } catch (e: JavascriptEvaluationException) {
            logger.error(e) { "Failed to run JavaScript" }
            "undefined"
        }
    }

    fun registerGlobalObject(key: String, value: Any) {
        globalObjects[key] = value
    }

    @Subscribe
    fun onResize(event: ResizeEvent) = view.resize(event.width.toLong(), event.height.toLong())

    fun isInputBlocked() = page?.blocksGameInput() ?: false

    override fun onDOMReady(frameId: Long, isMainFrame: Boolean, url: String?) {
        page?.onDocumentReady()
    }

    override fun onFailLoading(
        frameId: Long,
        isMainFrame: Boolean,
        url: String?,
        description: String?,
        errorDomain: String?,
        errorCode: Int
    ) {
        logger.error { "Ultralight failed loading '$url': $description ($errorCode @ $errorDomain)" }
    }

    override fun onFinishLoading(frameId: Long, isMainFrame: Boolean, url: String?) {
        page?.run {
            contextProvider.syncWithJavascript {
                contextProvider.registerObject(it.context, "_remote", this)
                for (global in globalObjects)
                    contextProvider.registerObject(it.context, global.key, global.value)
                this.onLoaded()
                runScript("if (typeof OnLoad !== 'undefined') { OnLoad(); }")

                synchronized(loadLock) {
                    finishedLoading = true
                    loadLock.notifyAll()
                }
            }
        }
    }

    override fun onBeginLoading(frameId: Long, isMainFrame: Boolean, url: String?) {
        synchronized(loadLock) {
            finishedLoading = false
        }
    }

    override fun onUpdateHistory() = Unit
    override fun onWindowObjectReady(frameId: Long, isMainFrame: Boolean, url: String?) = Unit

}